File: codes/ADTs/q6.java
public class DynamicStack {
    private int[] data;        // Array to store stack elements
    private int top;          // Index of the top element
    private int maxSize;     // Current maximum size of the stack

    /**
     * Initialize a new stack.
     * @param initialSize Initial size of the stack.
     */
    public DynamicStack(int initialSize) {
        this.data = new int[initialSize];
        this.top = -1;
        this.maxSize = initialSize;
    }

    /**
     * Check if the stack is empty.
     * @return true if the stack is empty, false otherwise.
     */
    public boolean isEmpty() {
        return top == -1;
    }

    /**
     * Resize the stack to a new maximum size.
     * @param newSize The new size for the stack.
     */
    private void resize(int newSize) {
        int[] newData = new int[newSize];
        // Copy existing elements to new array
        System.arraycopy(data, 0, newData, 0, maxSize);
        data = newData;
        maxSize = newSize;
        System.out.println("Resized stack to " + maxSize);
    }

    /**
     * Push an element onto the stack.
     * @param value The value to push.
     */
    public void push(int value) {
        if (top == maxSize - 1) { // If stack is full, resize
            resize(maxSize * 2);
        }
        data[++top] = value;
    }

    /**
     * Pop an element from the stack.
     * @return The popped value.
     * @throws IllegalStateException if the stack is empty.
     */
    public int pop() {
        if (isEmpty()) {
            throw new IllegalStateException("Error: Stack underflow");
        }
        return data[top--];
    }

    /**
     * Main method to demonstrate stack operations.
     */
    public static void main(String[] args) {
        DynamicStack stack = new DynamicStack(5); // Initial size of the stack

        // Push elements onto the stack
        for (int i = 1; i <= 10; i++) {
            stack.push(i * 10); // Pushing 10, 20, ..., 100
        }

        // Pop elements from the stack
        while (!stack.isEmpty()) {
            try {
                System.out.println("Popped value: " + stack.pop());
            } catch (IllegalStateException e) {
                System.out.println(e.getMessage());
                break;
            }
        }
    }
}

File: codes/ADTs/q6.c
// Time complexity of O(1) for the push and pop operations, and 
// A space complexity that grows dynamically with the number of elements in the stack. 

#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *data;        // Pointer to dynamically allocated array
    int top;          // Index of the top element
    int maxSize;     // Current maximum size of the stack
} Stack;

/**
 * Initialize a new stack.
 * @param initialSize Initial size of the stack.
 */
Stack* createStack(int initialSize) {
    Stack *stack = (Stack*)malloc(sizeof(Stack));
    stack->data = (int*)malloc(initialSize * sizeof(int));
    stack->top = -1;
    stack->maxSize = initialSize;
    return stack;
}

/**
 * Check if the stack is empty.
 * @return 1 if the stack is empty, 0 otherwise.
 */
int isEmpty(Stack *stack) {
    return stack->top == -1;
}

/**
 * Resize the stack to a new maximum size.
 * @param stack The stack to resize.
 * @param newSize The new size for the stack.
 */
void resize(Stack *stack, int newSize) {
    stack->data = (int*)realloc(stack->data, newSize * sizeof(int));
    if (stack->data == NULL) {
        fprintf(stderr, "Error: Unable to resize stack\n");
        exit(EXIT_FAILURE);
    }
    stack->maxSize = newSize;
}

/**
 * Push an element onto the stack.
 * @param stack The stack to push to.
 * @param value The value to push.
 * @return 1 if the push was successful, 0 if the stack is full.
 */
int push(Stack *stack, int value) {
    if (stack->top == stack->maxSize - 1) { // If stack is full, resize
        resize(stack, stack->maxSize * 2);
        printf("Resized stack to %d\n", stack->maxSize);
    }
    stack->data[++stack->top] = value;
    return 1;
}

/**
 * Pop an element from the stack.
 * @param stack The stack to pop from.
 * @param value Pointer to store the popped value.
 * @return 1 if the pop was successful, 0 if the stack is empty.
 */
int pop(Stack *stack, int *value) {
    if (isEmpty(stack)) {
        printf("Error: Stack underflow\n");
        return 0;
    }
    *value = stack->data[stack->top--];
    return 1;
}

/**
 * Free the memory allocated for the stack.
 * @param stack The stack to free.
 */
void freeStack(Stack *stack) {
    free(stack->data);
    free(stack);
}

int main() {
    Stack *stack = createStack(5); // Initial size of the stack

    // Push elements onto the stack
    for (int i = 1; i <= 10; i++) {
        push(stack, i * 10); // Pushing 10, 20, ..., 100
    }

    int value;
    // Pop elements from the stack
    while (pop(stack, &value)) {
        printf("Popped value: %d\n", value);
    }

    // Free allocated memory
    freeStack(stack);

    return 0;
}
